mne/beamformer/_dics.py

```diff
@@ -1,27 +1,587 @@
+    allowed_ori = [None, 'normal', 'max-power']
+    if pick_ori not in allowed_ori:
+        raise ValueError('"pick_ori" should be one of %s.' % allowed_ori)
+
+    # Leadfield rank and optional rank reduction
+    # (to deal with problems with complex eigenvalues within the computation
+    # of the optimal orientation when using pinv if the leadfield was only
+    # rank 2 (e.g., with the spherical headmodel of the phantom data),
+    # see gh-4568 and gh-4628.
+    if reduce_rank and not (pick_ori == 'max-power' and inversion == 'matrix'):
+        raise NotImplementedError(
+            'The computation of spatial filters with rank reduction using '
+            'reduce_rank=True is only implemented with pick_ori=="max-power" '
+            'and inversion="matrix".'
+        )
+
+    # Inversion mode
+    if inversion not in ['single', 'matrix']:
+        raise ValueError("The inversion parameter should be either 'single' "
+                         "or 'matrix'.")
+
+    frequencies = [np.mean(freq_bin) for freq_bin in csd.frequencies]
+    n_freqs = len(frequencies)
+    n_orient = forward['sol']['ncol'] // forward['nsource']
+
+    # Determine how to normalize the leadfield
+    if normalize_fwd:
+        if inversion == 'single':
+            fwd_norm = 'dipole'
+        else:
+            fwd_norm = 'vertex'
+    else:
+        fwd_norm = None  # No normalization
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori,
+        fwd_norm=fwd_norm,
+    )
+    csd_picks = [csd.ch_names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_data(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        if (pick_ori == 'max-power' and weight_norm == 'unit-noise-gain' and
+                inversion == 'matrix'):
+            Cm_inv_sq = Cm_inv.dot(Cm_inv)
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # XXX This should be de-duplicated with LCMV
```

Comment:
for next PR? if so open an issue

Reply:
I added this comment and yes, so I'll open it

Reply:
#5109